File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/javascript/05. Fast Power.js
/**
QUESTION:
Implement fast exponentiation pow(x, n) via binary exponentiation.

Approach:
Handle base case n = 0. Track sign of exponent; work with absolute n while repeatedly squaring base and multiplying the result when the current bit is set. If exponent was negative, invert the final result.

Time Complexity: O(log |n|)
Space Complexity: O(1)
*/

function myPow(x, n) {
    if (n === 0) return 1;
    let res = 1;
    let exponent = Math.abs(n);
    let base = x;
    while (exponent > 0) {
        if (exponent & 1) {
            res *= base;
        }
        base *= base;
        exponent >>= 1;
    }
    return n < 0 ? 1 / res : res;
}

module.exports = { myPow };

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/python/05. Fast Power.py
"""QUESTION:
Implement pow(x, n) via fast binary exponentiation.

Approach:
Handle n = 0 early. Track whether the exponent is negative, work with its absolute value, and repeatedly square the base while multiplying the result when the current bit is set. If n was negative, invert the final result.

Time Complexity: O(log |n|)
Space Complexity: O(1)
"""

def my_pow(x: float, n: int) -> float:
    if n == 0:
        return 1.0
    exponent = abs(n)
    base = x
    res = 1.0
    while exponent > 0:
        if exponent & 1:
            res *= base
        base *= base
        exponent >>= 1
    return 1.0 / res if n < 0 else res
